toptal/picasso
# fix: scroll to error while Form is rerendered

[SPT-1924](https://toptal-core.atlassian.net/browse/SPT-1924)

### Description

If `<Form />` is rerendered while submitting and there is an inline field validation error, the form will not scroll to the field with an error. The reason is that because we update state in those lines:
https://github.com/toptal/picasso/blob/d2f3e4de7af527ec13a0855c218c622159cbd90f/packages/picasso-forms/src/utils/scroll-to-error-decorator.ts#L41-L45
and form submit function is already created with "old" state:
https://github.com/toptal/picasso/blob/d2f3e4de7af527ec13a0855c218c622159cbd90f/packages/picasso-forms/src/utils/scroll-to-error-decorator.ts#L61-L65
To fix this issue, we'd want to move `state` object outside of decorator.

### How to test
- open: https://codesandbox.io/s/gallant-breeze-56y52?file=/src/App.tsx
- fill in the last name
- fill in the first name
- scroll to top
- submit form button
- **actual result**: we cannot see which field has errors
- **expected result**: form should scroll to field with error

You can also check `How to test` section in this PR: https://github.com/toptal/staff-portal/pull/4207 which I've already used alpha package with those changes

### Screenshots
**Before**
<video src="https://user-images.githubusercontent.com/4816942/134457621-d065a717-3d96-40b0-8c9d-421f0a649d92.mp4" />

**After**
<video src="https://user-images.githubusercontent.com/4816942/134457621-d065a717-3d96-40b0-8c9d-421f0a649d92.mp4" />

### Review

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>
